File: packages/android/src/main/kotlin/dev/tano/plugins/HapticsPlugin.kt
package dev.tano.plugins

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import dev.tano.bridge.TanoPlugin

/**
 * Native haptic feedback plugin for Tano on Android.
 *
 * Provides haptic feedback using [Vibrator] / [VibrationEffect] on Android.
 * On API 26+ uses VibrationEffect; on older devices falls back to the
 * deprecated `vibrate(long)` call.
 *
 * Supported methods: `impact`, `notification`, `selection`.
 *
 * Mirrors the iOS HapticsPlugin.
 *
 * @param context Android Context needed to access the Vibrator system service.
 */
class HapticsPlugin(private val context: Context) : TanoPlugin {

    override val name: String = "haptics"
    override val permissions: List<String> = emptyList()

    private val vibrator: Vibrator
        get() = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val manager = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            manager.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }

    // -- Routing --

    override suspend fun handle(method: String, params: Map<String, Any?>): Any? {
        return when (method) {
            "impact" -> impact(params)
            "notification" -> notification(params)
            "selection" -> selection()
            else -> throw IllegalArgumentException("Unknown haptics plugin method: $method")
        }
    }

    // -- impact --

    private fun impact(params: Map<String, Any?>): Map<String, Any?> {
        val style = params["style"] as? String ?: "medium"

        val durationMs = when (style) {
            "light" -> 20L
            "medium" -> 40L
            "heavy" -> 60L
            else -> throw IllegalArgumentException("Invalid value '$style' for parameter 'style'")
        }

        vibrate(durationMs)
        return mapOf("success" to true)
    }

    // -- notification --

    private fun notification(params: Map<String, Any?>): Map<String, Any?> {
        val type = params["type"] as? String ?: "success"

        val durationMs = when (type) {
            "success" -> 30L
            "warning" -> 50L
            "error" -> 80L
            else -> throw IllegalArgumentException("Invalid value '$type' for parameter 'type'")
        }

        vibrate(durationMs)
        return mapOf("success" to true)
    }

    // -- selection --

    private fun selection(): Map<String, Any?> {
        vibrate(10L)
        return mapOf("success" to true)
    }

    // -- Helpers --

    private fun vibrate(durationMs: Long) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(
                VibrationEffect.createOneShot(durationMs, VibrationEffect.DEFAULT_AMPLITUDE)
            )
        } else {
            @Suppress("DEPRECATION")
            vibrator.vibrate(durationMs)
        }
    }
}

File: packages/plugins/haptics/Sources/TanoPluginHaptics/HapticsPlugin.swift
import Foundation
import TanoBridge

#if canImport(UIKit)
import UIKit
#endif

/// Native haptic feedback plugin for Tano.
///
/// Provides haptic feedback using UIKit feedback generators on iOS.
/// On macOS (and in tests), methods succeed without triggering hardware.
///
/// Supported methods: `impact`, `notification`, `selection`.
public final class HapticsPlugin: TanoPlugin {

    // MARK: - TanoPlugin conformance

    public static let name = "haptics"
    public static let permissions: [String] = []

    public init() {}

    // MARK: - Routing

    public func handle(method: String, params: [String: Any]) async throws -> Any? {
        switch method {
        case "impact":
            return try await impact(params: params)
        case "notification":
            return try await notification(params: params)
        case "selection":
            return try await selection()
        default:
            throw HapticsPluginError.unknownMethod(method)
        }
    }

    // MARK: - impact

    private func impact(params: [String: Any]) async throws -> [String: Any] {
        let style = params["style"] as? String ?? "medium"

        #if canImport(UIKit) && !targetEnvironment(simulator) && !os(macOS)
        let feedbackStyle: UIImpactFeedbackGenerator.FeedbackStyle
        switch style {
        case "light":
            feedbackStyle = .light
        case "medium":
            feedbackStyle = .medium
        case "heavy":
            feedbackStyle = .heavy
        default:
            throw HapticsPluginError.invalidParameter("style", style)
        }
        await MainActor.run {
            let generator = UIImpactFeedbackGenerator(style: feedbackStyle)
            generator.impactOccurred()
        }
        #else
        // Validate style even on macOS / test environments
        guard ["light", "medium", "heavy"].contains(style) else {
            throw HapticsPluginError.invalidParameter("style", style)
        }
        #endif

        return ["success": true]
    }

    // MARK: - notification

    private func notification(params: [String: Any]) async throws -> [String: Any] {
        let type = params["type"] as? String ?? "success"

        #if canImport(UIKit) && !targetEnvironment(simulator) && !os(macOS)
        let feedbackType: UINotificationFeedbackGenerator.FeedbackType
        switch type {
        case "success":
            feedbackType = .success
        case "warning":
            feedbackType = .warning
        case "error":
            feedbackType = .error
        default:
            throw HapticsPluginError.invalidParameter("type", type)
        }
        await MainActor.run {
            let generator = UINotificationFeedbackGenerator()
            generator.notificationOccurred(feedbackType)
        }
        #else
        guard ["success", "warning", "error"].contains(type) else {
            throw HapticsPluginError.invalidParameter("type", type)
        }
        #endif

        return ["success": true]
    }

    // MARK: - selection

    private func selection() async -> [String: Any] {
        #if canImport(UIKit) && !targetEnvironment(simulator) && !os(macOS)
        await MainActor.run {
            let generator = UISelectionFeedbackGenerator()
            generator.selectionChanged()
        }
        #endif

        return ["success": true]
    }
}

// MARK: - Errors

public enum HapticsPluginError: Error, LocalizedError {
    case unknownMethod(String)
    case invalidParameter(String, String)

    public var errorDescription: String? {
        switch self {
        case .unknownMethod(let m):
            return "Unknown haptics plugin method: \(m)"
        case .invalidParameter(let param, let value):
            return "Invalid value '\(value)' for parameter '\(param)'"
        }
    }
}
